Acl/user management7.0

ryanrath opened the pull request:
## Description
Changes to provide support for managing Acls via the internal dashboard User Management tab. 

## Motivation and Context
While Acls work alongside the current Role system. It does not allow for the addition of, and subsequent management of, an new Acl via the XDMoD Internal Dashboard. That is now supported.

## Tests performed
Manual Tests were performed. They were as follows: 

**Create a New User:**
  - Login to the Internal Dashboard
  - Click the 'User Management tab.
  - Click the 'Existing Users' sub-tab
  - Click the 'Create & Manage Users' Toolbar button
  - Ensure that the 'New User' tab is selected
  - Fill in the following fields:
    - First Name
    - Last Name
    - E-Mail Address
    - Username
    - Map To ( the selection does not matter )
    - Acls:
      - Select the 'User' acl
      - Select any other Acls that are appropriate
    - Optional:
      - Set a different 'User Type' 
  - Click the 'Create User' button
  - Ensure that the process complete's without error. 
    - **Note:** If you are testing on a limited system, sendmail needs to be setup or the creation process with throw an error. But, the user account will still be setup correctly. Just refresh the page and you should see the newly created User account.
  - If the user account is not visible click the 'Refresh' toolbar button. The user account should now be visible.
  - Double click the table row that represents the newly created User account. 
  - Ensure that the 'Create & Manage Users' modal window has been shown. 
  - Ensure that the user that was clicked is pre-selected on the left side of the window.
  - Ensure that the 'Acl Assignment' portion of the modal is showing the correct selections.
  - Click the 'Actions' Toolbar menu item
  - Select the 'Disable This Account' menu item
  - Ensure that a confirmation pop-up is displayed asking to confirm if you want to disable access for the user.
  - Click 'Yes'
  - Ensure that the operation completes successfully.
  - Ensure that the user entry in the modal windows 'Existing Users' table is now colored red.
  - Click the 'Actions' Toolbar menu item
  - Select the 'Enable This Account' menu item
  - Ensure that a confirmation pop-up is displayed asking to confirm if you want to enable access for the user.
  - Click 'Yes'
  - Ensure that the operation completes successfully.
  - Ensure that the user entry in the modal windows 'Existing Users' tale is now black
  - Add or remove some additional acls
  - Click 'Save Changes'
  - Ensure that the operation completes successfully
  - Close the 'User Management' modal window.
  - Force refresh the browser
  - Select the 'Existing Users' sub-tab.
  - double click the same user
  - Ensure that all of the user's information is as it was pre-refresh

**Modify an existing User:**
  - Login to the Internal Dashboard
  - Click the 'User Management tab.
  - Click the 'Existing Users' sub-tab
  - Double click on an existing user in the table of users.
  - Ensure that the 'User Management' modal window is displayed
  - Ensure that the user that was originally clicked is visible and selected in the 'Existing Users' table.
  - Ensure that the 'User Details' of the selected user are as expected.
  - Add or remove some acls. 
  - Click 'Save Changes' 
  - Ensure that the operation completes successfully
  - Close the 'User Management' modal window
  - double click the modified user
  - Ensure that acls displayed are as expected.
  - Force Refresh the browser.
  - Select the 'Existing Users' sub-tab
  - double click the modified user.
  - Ensure that the 'User Management' modal is displayed.
  - Ensure that the selected user is both visible and selected in the 'Existing Users' table.
  - Ensure that all of the information in the 'User Details' panel is as expected.

## Types of changes
<!--- What types of changes does your code introduce? Put an `x` in all the boxes that apply: -->
- [X] Bug fix (non-breaking change which fixes an issue)
- [X] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:
<!--- Go over all the following points, and put an `x` in all the boxes that apply. -->
<!--- If you're unsure about any of these, don't hesitate to ask. We're here to help! -->
- [X] My code follows the code style of this project as found in the **CONTRIBUTING** document.
- [ ] I have added tests to cover my changes.
- [X] All new and existing tests passed.

smgallo commented on html/controllers/sab_user/enum_tg_users.php:
```diff
@@ -38,13 +40,13 @@
 
 if (
-    $user->getActiveRole()->getIdentifier() == ROLE_ID_CAMPUS_CHAMPION
+   $user->hasAcl(ROLE_ID_CAMPUS_CHAMPION)
```
I don't see a campus champion role in the `roles.json` file in this PR.

ryanrath replied:
Agreed, this was just a straight replace of the pattern
```php 
$user->getActiveRole()->getIdentifier() === SOME_HARD_CODED_ROLE;
```
with:
```php
$user->hasAcl(SOME_HARD_CODED_ROLE);
```
If the code was there before I didn't want to remove it and possibly break something that utilizes that logic / functionality / data down the line.

smgallo replied:
ok

smgallo commented on html/controllers/ui_data/summary3.php:
```diff
@@ -64,8 +67,13 @@
-
-    $summaryCharts = $logged_in_user->getMostPrivilegedRole()->getSummaryCharts();
+    $mostPrivilegedAcl = Acls::getMostPrivilegedAcl($logged_in_user);
+    $summaryCharts = array_map(
```
This function encodes the summary charts and then loops through them at least twice unencoding the JSON, setting a value, and re-encoding. Why not just encode them at the end?

ryanrath replied:
Yeah, again I just wanted to sub out the deprecated function usage w/ the minimal amount of new to make it work. I didn't want to do any optimizations until I knew everything was still copacetic. But, it can most definitely be simplified. I'll put that in.

smgallo commented on html/controllers/user_admin/enum_roles.php:
```diff
@@ -1,28 +1,39 @@
+   $roles = $xda->enumerateAcls();
+
+   try {
+       $multipleServiceProviders = \xd_utilities\getConfiguration('features', 'multiple_service_providers') === 'on';
```
I'd prefer to see statements like this use parens to make it easier to read. E.g., `$value = ( func() === $otherValue);`

smgallo commented on html/controllers/user_admin/update_user.php:
```diff
@@ -104,8 +100,22 @@
+           $acl = Acls::getAclByName($aclName);
+           $user_to_update->addAcl($acl);
+
+           if (count($centers) > 0) {
```
Assuming that $centers is guaranteed to be an array you don't need the `if` statement here.

smgallo commented on html/controllers/user_admin/update_user.php:
```diff
@@ -104,8 +100,22 @@
+    }
 
-	if (isset($_POST['roles'])) {
+	/*if (isset($_POST['roles'])) {
```
Remove commented out code.

smgallo commented on html/internal_dashboard/js/admin_panel/SectionNewUser.js:
```diff
@@ -491,21 +491,15 @@ XDMoD.CreateUser = Ext.extend(Ext.form.FormPanel, {
+                var missingMappedUser = cmbUserMapping.getValue().length === 0;
 
-                if (!newUserRoleGrid.areRolesSpecified()) {
+                if (!acls.length > 0) {
```
Why not just `if ( acls.length <= 0)` ?

ryanrath replied:
sure

smgallo commented on html/internal_dashboard/js/admin_panel/SectionNewUser.js:
```diff
@@ -534,6 +525,13 @@ XDMoD.CreateUser = Ext.extend(Ext.form.FormPanel, {
+                var populatedAcls = {};
+                // populate centers for any of the selected acls that have them
+                for ( var i = 0; i < acls.length; i++) {
+                    var acl = acls[i];
```
Are these acl objects or ids?

smgallo commented:
@ryanrath This is set to merge into v7.0 but it should be v7.1

ryanrath commented on configuration/roles.json:
```diff
@@ -349,39 +353,59 @@
         "cd": {
             "display": "Center Director",
-            "type": "feature",
+            "type": "feature_requires_center",
```
so the intended purpose is to indicate that these acls require, when the install is using multiple service providers, additional information to function properly. What this translates to is a record being recorded in the user_acl_group_by_parameters table w/ a value of the service providers ( center ) id that should be associated with a user / acl / group by ( provider in this case ). 

But now that I've had a bit to look it over, a better way to indicate this requirement might be to do something like: 
```json
{
  "cd": {
    "display": "Center Director",
    "type": {
      "name": "feature",
      "requires_center": true
    }
  }
}
```

smgallo replied:
Or something more generic like this so multiple items could be required. Then the question is, where do the definition of the require values come from.
```
{
  "cd": {
    "display": "Center Director",
    "type": {
      "name": "feature",
      "requires": [ "center" ]
    }
  }
}

ryanrath replied:
Ooooo, I like that. Ok, I'll work on updating the code / db to work with the new config layout.

ryanrath commented:
Merge target changed to 7.1. Working on some disabled menus issues at the moment.

smgallo commented on classes/Models/Services/Acls.php:
```diff
@@ -780,4 +776,86 @@ public static function getDescriptorParamValues(XDUser $user, $aclName, $groupBy
+     */
+    public static function getMostPrivilegedAcl(XDUser $user, $moduleName = DEFAULT_MODULE_NAME, $aclHierarchyName = 'acl_hierarchy')
+    {
+        if (null == $user->getUserID()) {
```
Use === here because null == 0 is true

smgallo commented on bin/acl-config:
```diff
@@ -864,43 +865,40 @@ SQL;
+    } elseif ($count === 1) {
+        $log->info("[SUCCESS] Processed Acl Hierarchy: $id");
+    } else {
+        $log->info("[FAIL] Inserted more than one Acl Hierarchy Record for: $id");
```
This should still be wrapped in a transaction.

smgallo commented on classes/XDUser.php:
```diff
@@ -3194,4 +3177,68 @@ public static function getUserByUserName($username)
+    public function addAclOrganization($aclName, $organizationId)
+    {
+        if (empty($this->_id)) {
+            throw new \Exception("This user must be saved prior to calling this function.");
```
Change "calling this function" to "calling " . `__FUNCTION__` for clarity.

smgallo commented on classes/XDUser.php:
```diff
@@ -3194,4 +3177,55 @@ public static function getUserByUserName($username)
+        $acl = Acls::getAclByName($aclName);
+
+        if ( null == $acl) {
+            throw new Exception("Unable to retrieve acl for: $aclName");
```
There are still several places where `\Exception` is being used and also `Exception`. These should be consistent. If you want to do this in another PR that's fine.

chakrabortyr commented on bin/acl-config:
```diff
@@ -243,7 +244,7 @@ SQL;
 
             if (null === $hierarchy['name']) {
-                throw new Exception("Malformed hierarchy entry for module $moduleName. No name property.");
+                throw new \Exception("Malformed hierarchy entry for module $moduleName. No name property.");
```
This may should be just Exception without the backslash as per [this pr](https://github.com/ubccr/xdmod/pull/299).

chakrabortyr commented on classes/Models/Services/Acls.php:
```diff
@@ -442,15 +445,8 @@ public static function getDisabledMenus(XDUser $user, array $realms)
-            }
             if ($row['id'] != null) {
-                $results[$name] [] = array(
+                $results[] = array(
```
What's the behavior when you remove the index from the initial array? Does this always produce the desired result?

ryanrath replied:
The behavior is that instead of having an associative array indexed by acl.name we now have a non-associative array that contains an record for each row that has a non-null id. I do need to update the comments as they no longer make sense, but this was changed so that the value of the `CCR.xdmod.ui.disabledMenus` property matched what it was before.

smgallo commented:
Looks like there are some still some issues with the tests.

jpwhite4 commented on bin/acl-config:
```diff
@@ -914,6 +941,16 @@ SQL;
+        $log->warning("[FAIL] Attempt to commit hierarchies for Acl [$acl] failed.");
+        $rolledBack = $db->rollBack();
+
+        // We throw an Exception so that we do not leave the db in an inconsistent state.
```
This comment is at odds with the error message - if throwing the exception stops the database being left in an incosistent state then why does the exception claim that the database may be inconsistent?

jpwhite4 commented:
Please can you add some tests for the new functionality. I suggest using the integration test harness to call the new/modified internal dashboard API calls to add/remove/modify some user accounts.

jpwhite4 commented:
The "Public User" also incorrectly appears in the New User dialog box too.

Also I managed to get it to display this message:

![image](https://user-images.githubusercontent.com/5342179/31635326-539dccc6-b294-11e7-99e0-dbbf6f514e1b.png)

but it still create an entry in the database for the user anyway. This probably needs to not add the rows to the database in this case.

Even better than that the user is undeletable now.

jpwhite4 commented:
Also the Public User shows up in the Existing Users grid - this too should be hidden

chakrabortyr commented:
User tests are still failing

ryanrath commented:
@jpwhite4 re: Public User showing in the existing users grid. I can certainly make that happen. For what it's worth the 'Public User' is visible in prod when you go to 'Create / Manage Users' and select 'Internal Users'. Would this be considered a bug? Or is this behavior desired? I just want to make sure before I start making changes.

ryanrath commented:
@jpwhite4 for the user creation exception you encountered, did you provide a username / mapped user and were there any acls assigned ( all I can see is that center director & center staff were not assigned )?

ryanrath commented:
Please note that there are some tests that were commented out during development. I'll be committing them shortly.

jpwhite4 commented on configuration/roles.json:
```diff
@@ -361,15 +361,15 @@
         "cd": {
             "display": "Center Director",
-            "type": "feature",
+            "extends": "default",
```
Why did you change this?

jpwhite4 commented on configuration/roles.json:
```diff
@@ -227,7 +227,7 @@
                                 },
                                 "group_by": "pi",
                                 "has_std_err": true,
-                                "id": 0.00000000000004,
```
Why did this change?

smgallo commented:
In XDUser.php there are several exceptions being thrown as `Exception` and others as '\Exception`. Since this is not a namespaced file it doesn't matter but it should be consistent.

ryanrath commented:
@jpwhite4 I've removed `Public User` / the `pub` role from the internal_dashboard -> User Management tab. I've also updated the ControllerTests so that they take this removal into account.
